src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,51 +29,111 @@ class DictionarySegment : public BaseSegment {
+    } else if (dictionary_size <= std::numeric_limits<uint32_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint32_t>>(value_size);
+    }
+    DebugAssert(_attribute_vector, "Too many unique values");
+
+    // Second pass: Fill the attribute vector
+    for (size_t value_index = 0; value_index < value_size; value_index++) {
+      // The set that we used to construct the _dictionary vector is ordered by the "less" comparator,
+      // so the constructed _dictionary is already sorted. Thus, we can find the index for each
+      // value in the dictionary in O(log(n)).
+      const auto& value = values[value_index];
+      const auto it = std::lower_bound(_dictionary->cbegin(), _dictionary->cend(), value);
+      size_t dic_index = std::distance(_dictionary->cbegin(), it);
+      _attribute_vector->set(value_index, static_cast<ValueID>(dic_index));
+    }
+  }
 
   // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
   // the DictionarySegment in this file. Replace the method signatures with actual implementations.
 
   // return the value at a certain position. If you want to write efficient operators, back off!
-  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override;
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override {
+    return (*_dictionary)[_attribute_vector->get(chunk_offset)];
+  }
 
   // return the value at a certain position.
-  T get(const size_t chunk_offset) const;
+  T get(const size_t chunk_offset) const { return (*_dictionary)[_attribute_vector->get(chunk_offset)]; }
 
   // dictionary segments are immutable
-  void append(const AllTypeVariant&) override;
+  void append(const AllTypeVariant&) override {
+    throw std::runtime_error("append() called on immutable dictionary segment");
+  }
 
   // returns an underlying dictionary
-  std::shared_ptr<const std::vector<T>> dictionary() const;
+  std::shared_ptr<const std::vector<T>> dictionary() const { return _dictionary; }
 
   // returns an underlying data structure
-  std::shared_ptr<const BaseAttributeVector> attribute_vector() const;
+  std::shared_ptr<const BaseAttributeVector> attribute_vector() const { return _attribute_vector; }
 
   // return the value represented by a given ValueID
-  const T& value_by_value_id(ValueID value_id) const;
+  const T& value_by_value_id(ValueID value_id) const { return _dictionary->at(value_id); }
 
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    auto it = std::lower_bound(_dictionary->cbegin(), _dictionary->cend(), value);
+
+    if (it != _dictionary->cend()) {
+      return static_cast<ValueID>(std::distance(_dictionary->cbegin(), it));
```

Comment:
you can use ValueID(id) constructor instead of cast

Reply:
This gave us a warning for narrowing when building the tests.

When we have narrowing we decided to use static_cast